### PR TITLE
Provide option to pass module instance to _load_state_dict_pre_hooks.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62242
* __->__ #62070

We have a custom Tensor:
https://github.com/pytorch/pytorch/blob/master/torch/distributed/_sharded_tensor/api.py#L67,
which doesn't show up in state_dict for the module. This was resolved by
using the _register_state_dict_hook:
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L1196
to parse and add custom tensors to state_dict.

However, the problem is during load time  _register_load_state_dict_pre_hook:
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L1272,
does not pass in the module instance and as a result, a ShardedTensor in the
state_dict cannot be appropriately added to a module at load time.

To resolve this issue, in this PR I've enhanced this hook to support two
variations, one which passes in the module instance (for the problem described
above) and one is the previous version for BC reasons.

Differential Revision: [D29867142](https://our.internmc.facebook.com/intern/diff/D29867142/)

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -1312,7 +1314,13 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
                 :meth:`~torch.nn.Module.load_state_dict`
         """
         for hook in self._load_state_dict_pre_hooks.values():
-            hook(state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs)
+            n_params = len(signature(hook).parameters)
+            if n_params == 8:
+                # Pass in module instance as well.
+                hook(state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs, self)
```
Comment (pritamdamania87) [inline torch/nn/modules/module.py:1320]: This is a bit hacky, another option I had in mind was to have an additional option in _register_load_state_dict_pre_hook something like `with_module` and store that option with the hook. Then during load time, based on that option we either pass in the module or we don't for the hook.

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -1269,14 +1269,20 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
...
+
+        Arguments:
+            hook (Callable): Callable hook that will be invoked before
+                loading the state dict.
+            with_module (bool, optional): Whether or not to pass the module
+                instance to the hook.
         """
         handle = hooks.RemovableHandle(self._load_state_dict_pre_hooks)
-        self._load_state_dict_pre_hooks[handle.id] = hook
+        self._load_state_dict_pre_hooks[handle.id] = (hook, with_module)
```
Comment (albanD) [inline torch/nn/modules/module.py:1285]: can't you use partial here to just update the hook functions?

Comment (pritamdamania87) [reply torch/nn/modules/module.py:1285]: Thinking about this a bit more, the problem is that this is a custom user hook and we don't have any easy way to know the name of the last parameter (unless we use inspect). If we use partial, we probably need to do something like `hook  = partial(hook, <last_param_name>='self')`

Comment (albanD) [reply torch/nn/modules/module.py:1285]: Well, you're writing a brand new API, you can make the module param the very first positional argument ;)

File: test/test_nn.py [hunk 3]
```diff
@@ -17980,6 +17980,62 @@ def test_pickle_softsign(self):
...
+
+        def hook_with_module(module, state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs):
+            self.assertEqual(m_state_dict, state_dict)
+            self.assertTrue(m_load is module)
+            pass
+
+        m_load._register_load_state_dict_pre_hook(hook_without_module)
+        m_load._register_load_state_dict_pre_hook(hook_with_module, True)
+
+        m_load.load_state_dict(m_state_dict)
```
Comment (jbschlosser) [inline test/test_nn.py:18004]: Should probably add some logic verifying that the hooks were called

File: test/test_nn.py [hunk 4]
```diff
@@ -17980,6 +17980,62 @@ def test_pickle_softsign(self):
...
+                assert [] == missing_keys
+                assert strict
+                assert self is module
+                pass
+
+        m = MyModule()
+        state_dict = m.state_dict()
+        m._register_load_state_dict_pre_hook(m.my_pre_load_hook)
+        m._register_load_state_dict_pre_hook(m.my_pre_load_hook_with_module, True)
+        m.load_state_dict(state_dict)
```
Comment (jbschlosser) [inline test/test_nn.py:18037]: Same as above: validate hooks are called

File: test/test_nn.py [hunk 5]
```diff
@@ -17980,6 +17980,62 @@ def test_pickle_softsign(self):
...
+            self.assertEqual(m_state_dict, state_dict)
+            self.assertTrue(m_load is module)
+            pass
+
+        m_load._register_load_state_dict_pre_hook(hook_without_module)
+        m_load._register_load_state_dict_pre_hook(hook_with_module, True)
+
+        m_load.load_state_dict(m_state_dict)
+
+        # Test with module instance method as hook
```
Comment (jbschlosser) [inline test/test_nn.py:18006]: nit: for readability, this could be split into a separate test

File: test/test_nn.py [hunk 6]
```diff
@@ -17980,6 +17980,62 @@ def test_pickle_softsign(self):
...
+            ):
+                assert [] == error_msgs
+                assert [] == unexpected_keys
+                assert [] == missing_keys
+                assert strict
+                assert self is module
+                pass
+
+        m = MyModule()
+        state_dict = m.state_dict()
```
Comment (jbschlosser) [inline test/test_nn.py:18034]: nit: `state_dict` is empty here- is it worth testing with a non-empty one by adding a parameter / buffer / submodule to `MyModule`?

**Review by jbschlosser — COMMENTED** [review]
Change looks good! Couple minor comments on testing below

**Review by jbschlosser — APPROVED** [review]
LGTM!

**Pull request closed without merging**